gematria/datasets/basic_block_utils.cc

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gematria/datasets/basic_block_utils.h"
+
+#include <map>
+
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86RegisterInfo.h"
+#include "llvm/MC/MCInstrInfo.h"
+#include "llvm/MC/MCRegisterInfo.h"
+
+using namespace llvm;
+
+namespace gematria {
+
+unsigned getSuperRegister(unsigned OriginalRegister,
+                          const MCRegisterInfo &RegisterInfo) {
+  unsigned SuperRegister = OriginalRegister;
+  for (MCPhysReg CurrentSuperRegister :
+       RegisterInfo.superregs_inclusive(OriginalRegister)) {
+    SuperRegister = CurrentSuperRegister;
+  }
+  // Only return super registers for GPRs. Since mainly GPRs will be used for
+  // addressing, redefining other aliasing registers (like vector registers)
+  // does not matter as much.
```

Comment:
For others (in particular for XMMs), this logic is fine, but we might want to post-process it afterwards, and if there are overlapping vector registers, take just the largest one.

Reply:
That's a good point. I'll look into doing that at some point. My main concern with immediately bumping all the vector registers too was that it is super easy to end up registers that aren't supported by the current CPU without looking at what is supported by the target (which is reasonably complicated logic). This works for now and should be adjustable in the future.